xla/service/gpu/runtime/nccl_all_to_all_thunk.cc

```diff
@@ -99,22 +98,37 @@ absl::Status NcclAllToAllStartThunk::RunNcclCollective(
       std::vector<DeviceBufferPair> device_buffers,
       ConvertToDeviceBuffers(params, buffers_,
                              config_.config.operand_element_type));
+
+  TF_ASSIGN_OR_RETURN(const int64_t current_id,
+                      GetCurrentId(params.collective_params, config_));
+
+  TF_ASSIGN_OR_RETURN(int32_t num_participants, nccl_api()->CommCount(comm_wrapper.comm_handle));
+  for (int64_t id = 0; id < num_participants; ++id) {
+    TF_RETURN_IF_ERROR(recv_ptr_map_.InitializeId(id, current_id));
+    TF_RETURN_IF_ERROR(recv_ptr_map_.InitializeId(current_id, id));
+  }
+
+  bool use_memcpy = comm_wrapper.is_local && p2p_memcpy_enabled_;
+  
   return xla::gpu::RunAllToAll(nccl_api(), config_.has_split_dimension,
                                device_buffers, stream,
-                               comm_wrapper.comm_handle);
+                               comm_wrapper.comm_handle, current_id, use_memcpy, recv_ptr_map_);
 }
 
 absl::Status RunAllToAll(NcclApi* nccl_api, bool has_split_dimension,
```

Comment:
This is implemented in two modes which do not share much code. Can we outline that into two functions and dispatch here?

Reply:
broke to two functions.